### PR TITLE
[tunix] Preserve fractional weighted-gradient denominators

### PR DESCRIPTION
**Pull request by ethannnnnn**
## Motivation

`LossOutput` carries an unreduced numerator and its denominator so gradient
accumulation can preserve the weighting of every token or example. Tunix now
accumulates numerator gradients using the global denominator, but
`GradientAccumulator.get()` still clamps that denominator to at least one.
That under-scales valid fractional weights. The depth-one fast path also
divides directly by zero for an empty weighted batch.

## Scope

This PR changes only the remaining denominator edge cases in the generic SFT
trainer and its focused tests:

- preserve every positive fractional denominator exactly;
- return zero gradients when the accumulated denominator is zero;
- apply the same rule to the depth-one fast path; and
- cover fractional and zero denominators in both execution paths.

It does not add diffusion contracts, change metric logging or reduction,
introduce an SFT objective, or alter RL behavior.

## Design

For microbatches `i`, Tunix already computes

```text
sum_i grad(numerator_i) / sum_i denominator_i
```

This PR replaces `1 / max(sum_i denominator_i, 1)` with a zero-safe reciprocal:
positive denominators retain their exact value, while a zero total denominator
produces an all-zero gradient tree without evaluating a divide-by-zero path.
Legacy scalar losses still contribute denominator `1`, so their behavior is
unchanged.

## Compatibility

The behavior changes only when a loss function returns `utils.LossOutput`.
Plain scalar losses retain the existing gradient-accumulation semantics. The
public `LossOutput` and `WeightedMetric` data contracts are unchanged.

## Tests

```text
python -m pytest -q tests/sft/peft_trainer_test.py
```

The guarded publisher also runs changed-line Pyink, scoped error/fatal Pylint,
Python compilation, and `git diff --check` on the two changed files.

## Follow-Ups

Global reduction of weighted train and evaluation metrics is intentionally a
separate review. Diffusion training and RL integrations remain later layers.

Design document:
https://docs.google.com/document/d/1Xe-98ScS2RSH29AdhTdc9tO4WHQG5wCbFkJA3gmIUZQ/edit

### CODE REVIEW

File: tunix/sft/peft_trainer.py [hunk 1]
```diff
@@ -167,6 +167,11 @@ def _calculate_global_batch_size(train_example: Any) -> int:
   )
 
 
+def _zero_safe_reciprocal(denom: jax.Array) -> jax.Array:
+  safe_denom = jnp.where(denom == 0, jnp.ones_like(denom), denom)
+  return jnp.where(denom == 0, jnp.zeros_like(denom), 1.0 / safe_denom)
```
Comment (wang2yn84) [inline tunix/sft/peft_trainer.py]: nit, no need for safe_denom, we already have conditional check here, so return jnp.where(denom == 0, jnp.zeros_like(denom), 1.0 /denom) should be fine.

Comment (ethannnnnn) [reply tunix/sft/peft_trainer.py]: Done. `_zero_safe_reciprocal` now returns the zero-guarded `jnp.where` expression directly, without the intermediate `safe_denom`. The fractional- and zero-denominator tests still cover both the depth-one and accumulator-backed paths, and all 70 `PeftTrainer` tests pass.
